ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,411 @@
 
 #include "bcmath.h"
 #include "private.h"
+#include "convert.h"
 #include <stdbool.h>
 #include <stddef.h>
 #include <string.h>
 #include "zend_alloc.h"
 
+static const BC_VECTOR POW_10_LUT[9] = {
+	1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000
+};
 
-/* Some utility routines for the divide:  First a one digit multiply.
-   NUM (with SIZE digits) is multiplied by DIGIT and the result is
-   placed into RESULT.  It is written so that NUM and RESULT can be
-   the same pointers.  */
-
-static void _one_mult(unsigned char *num, size_t size, int digit, unsigned char *result)
+/*
+ * Use this function when the number of arrays in n2 is 1.
+ * In this case, there is no need to add back.
+ */
+static inline void bc_fast_div(BC_VECTOR *n1_vector, size_t n1_arr_size, BC_VECTOR n2_vector, BC_VECTOR *quot_vector, size_t quot_arr_size)
 {
-	size_t carry, value;
-	unsigned char *nptr, *rptr;
+	size_t n1_top_index = n1_arr_size - 1;
+	size_t quot_top_index = quot_arr_size - 1;
+	for (size_t i = 0; i < quot_arr_size - 1; i++) {
+		if (n1_vector[n1_top_index - i] < n2_vector) {
+			quot_vector[quot_top_index - i] = 0;
+			n1_vector[n1_top_index - i - 1] += n1_vector[n1_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+			continue;
+		}
+		quot_vector[quot_top_index - i] = n1_vector[n1_top_index - i] / n2_vector;
+		n1_vector[n1_top_index - i] -= n2_vector * quot_vector[quot_top_index - i];
+		n1_vector[n1_top_index - i - 1] += n1_vector[n1_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
```

Comment:
Added comment.

Reply:
> So if I understand correctly, the n1_vector[n1_top_index - i - 1] value may become bigger than BC_VECTOR_BOUNDARY_NUM after this, but since it's a single for loop this will not cause an overflow and we'll correct for it later by carrying the too large values similarly to how we do it for the multiplication code? (this is the first function I'm reviewing so maybe this question answers itself later)

Yes, that's right